services/statestorage/layered_state.go

```diff
@@ -0,0 +1,107 @@
+package statestorage
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/orbs-network/orbs-network-go/instrumentation/log"
+	"github.com/orbs-network/orbs-network-go/services/statestorage/adapter"
+	"github.com/orbs-network/orbs-spec/types/go/primitives"
+	"github.com/orbs-network/orbs-spec/types/go/protocol"
+	"github.com/pkg/errors"
+)
+
+type stateIncrement struct {
+	diff       adapter.ChainState
+	merkleRoot primitives.MerkleSha256
+	height     primitives.BlockHeight
+	ts         primitives.TimestampNano
+}
+
+type layeredState struct {
+	persistence  adapter.StatePersistence
+	maxLayers    int
+	incCache     []*stateIncrement
+	recentHeight primitives.BlockHeight
+	pHeight      primitives.BlockHeight
+	pRoot        primitives.MerkleSha256
+	pTs          primitives.TimestampNano
+}
+
+func newLayeredState(p adapter.StatePersistence, maxLayers int) *layeredState {
+	pHeight, pTs, pRoot, err := p.ReadMetadata()
+	if err != nil {
+		panic("could not load state metadata")
+	}
+	result := &layeredState{
+		persistence:  p,
+		maxLayers:    maxLayers,
+		recentHeight: pHeight,
+		pHeight:      pHeight,
+		pTs:          pTs,
+		pRoot:        pRoot,
+	}
+
+	return result
+}
+
+func (ls *layeredState) Write(height primitives.BlockHeight, ts primitives.TimestampNano, root primitives.MerkleSha256, diff adapter.ChainState) error {
+	ls.incCache = append(ls.incCache, &stateIncrement{
```

Comment:
is this thread safe?

Reply:
this object is not thread safe. everything here is called under a mutex from the call site